gc/base/Evacuator.cpp

```diff
@@ -0,0 +1,1438 @@
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "CollectorLanguageInterface.hpp"
+#include "EnvironmentStandard.hpp"
+#include "Evacuator.hpp"
+#include "EvacuatorController.hpp"
+#include "EvacuatorDelegate.hpp"
+#include "ForwardedHeader.hpp"
+#include "GCExtensionsBase.hpp"
+#include "IndexableObjectScanner.hpp"
+#include "Math.hpp"
+#include "MemcheckWrapper.hpp"
+#include "ObjectModel.hpp"
+#include "ScavengerStats.hpp"
+#include "SlotObject.hpp"
+#include "SublistFragment.hpp"
+
+extern "C" {
+	uintptr_t allocateMemoryForSublistFragment(void *vmThreadRawPtr, J9VMGC_SublistFragment *fragmentPrimitive);
+}
+
+MM_Evacuator *
+MM_Evacuator::newInstance(uintptr_t workerIndex, MM_EvacuatorController *controller, GC_ObjectModel *objectModel, MM_Forge *forge)
+{
+	MM_Evacuator *evacuator = (MM_Evacuator *)forge->allocate(sizeof(MM_Evacuator), OMR::GC::AllocationCategory::FIXED, OMR_GET_CALLSITE());
+	if(NULL != evacuator) {
+		new(evacuator) MM_Evacuator(workerIndex, controller, objectModel, forge);
+		if (!evacuator->initialize()) {
+			evacuator->kill();
+			evacuator = NULL;
+		}
+	}
+	return evacuator;
+}
+
+void
+MM_Evacuator::kill()
+{
+	tearDown();
+	_forge->free(this);
+}
+
+bool
+MM_Evacuator::initialize()
+{
+	/* initialize the evacuator mutex */
+	if (0 != omrthread_monitor_init_with_name(&_mutex, 0, "MM_Evacuator::_mutex")) {
+		return false;
+	}
+
+	/* initialize the delegate */
+	if (!_delegate.initialize(this, _forge, _controller)) {
+		return false;
+	}
+
+	return true;
+}
+
+void
+MM_Evacuator::tearDown()
+{
+	/* tear down delegate */
+	_delegate.tearDown();
+
+	/* tear down mutex */
+	if (NULL != _mutex) {
+		omrthread_monitor_destroy(_mutex);
+		_mutex = NULL;
+	}
+
+	/* free forge memory bound to arrays instantiated in constructor */
+	_forge->free(_stackBottom);
+	_forge->free(_copyspace);
+	_forge->free(_whiteList);
+
+	/* free the freelist */
+	_freeList.reset();
+}
+
+/**
+ * Per gc cycle setup. This binds the evacuator instance to a gc worker thread for the duration of the cycle.
+ *
+ * @param env the environment for the gc worker thread that is bound to this evacuator instance
+ */
+void
+MM_Evacuator::bindWorkerThread(MM_EnvironmentStandard *env)
+{
+	omrthread_monitor_enter(_mutex);
+
+	/* bind evacuator and delegate to executing gc thread */
+	_env = env;
+	_env->setEvacuator(this);
+	_delegate.cycleStart();
+
+	/* clear worker gc stats */
+	_stats = &_env->_scavengerStats;
+	_stats->clear(true);
+	_stats->_gcCount = _env->getExtensions()->scavengerStats._gcCount;
+
+	/* Reset the local remembered set fragment */
+	_env->_scavengerRememberedSet.count = 0;
```

Comment:
Is this the remembered set cache that is used during the GC for the current thread?

Reply:
It is.